### PR TITLE
agent: Add agent shutdown test

### PR DESCRIPTION
**Pull request by jodh-intel**
Run this test before the tracing test since agent tracing requires the agent
be able to shutdown gracefully. As such, ensure the shutdown is working
before moving on to testing tracing.

Fixes: #3494.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

File: tracing/test-agent-shutdown.sh [hunk 1]
```diff
@@ -0,0 +1,1341 @@
...
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../.ci/lib.sh"
+
+RUNTIME=${RUNTIME:-"io.containerd.kata.v2"}
+
+# Kata always uses this value
+EXPECTED_VSOCK_PORT="1024"
+
+DOCKER_IMAGE=${DOCKER_IMAGE:-"busybox"}
+CTR_DOCKER_IMAGE=${CTR_DOCKER_IMAGE:-"docker.io/library/busybox:latest"}
```
Comment (devimc) [inline tracing/test-agent-shutdown.sh:52]: please use a different registry , i.e public.ecr.aws/ubuntu/ubuntu

File: tracing/test-agent-shutdown.sh [hunk 2]
```diff
@@ -0,0 +1,1341 @@
...
+
+	for i in $(seq "$max")
+	do
+		{ sudo kill -0 "$agent_pid"; ret=$?; } || true
+
+		[ "$ret" -ne 0 ] && agent_ended="true" && break
+
+		sleep 0.2
+
+		i=$((i+1))
```
Comment (devimc) [inline tracing/test-agent-shutdown.sh:1149]: is this really needed ? I don't think so.. :thinking:

File: tracing/test-agent-shutdown.sh [hunk 3]
```diff
@@ -0,0 +1,1341 @@
...
+
+	addresses=$(ss -HO --vsock |\
+		awk '$2 ~ /^ESTAB$/ {print $6}')
+
+	[ -z "$addresses" ] && return 1
+
+	local expected_count=1
+
+	local count
+	count=$(echo "$addresses"|wc -l || true)
```
Comment (devimc) [inline tracing/test-agent-shutdown.sh:893]: uhmm, even when the string is empty, wc returns 1 :thinking: 

```
$ echo "" | wc -l
1
```

File: tracing/test-agent-shutdown.sh [hunk 4]
```diff
@@ -0,0 +1,1341 @@
...
+
+	setup_containerd
+}
+
+start_local_agent()
+{
+	local log_file="${1:-}"
+	[ -z "$log_file" ] && die "need agent log file"
+
+	[ -x "$agent_binary" ] || (cd "$agent_dir" && make)
```
Comment (devimc) [inline tracing/test-agent-shutdown.sh:691]: `make -C ${agent_dir}` looks better :wink:

File: tracing/test-agent-shutdown.sh [hunk 5]
```diff
@@ -0,0 +1,1341 @@
...
+
+	addresses=$(ss -HO --vsock |\
+		awk '$2 ~ /^ESTAB$/ {print $6}')
+
+	[ -z "$addresses" ] && die "no VSOCK connections found"
+
+	local expected_count=1
+
+	local count
+	count=$(echo "$addresses"|wc -l || true)
```
Comment (devimc) [inline tracing/test-agent-shutdown.sh:922]: same problem with wc

**Review by devimc — APPROVED** [review]
thanks @jodh-intel - I've left some comments

File: tracing/test-agent-shutdown.sh (hunk 1 above)
Comment (jodh-intel) [reply tracing/test-agent-shutdown.sh]: Fixed.

File: tracing/test-agent-shutdown.sh (hunk 2 above)
Comment (jodh-intel) [reply tracing/test-agent-shutdown.sh]: Good catch - fixed.

File: tracing/test-agent-shutdown.sh (hunk 3 above)
Comment (jodh-intel) [reply tracing/test-agent-shutdown.sh:893]: True, but the code has already checked to make sure `$addresses` is not empty.

File: tracing/test-agent-shutdown.sh (hunk 4 above)
Comment (jodh-intel) [reply tracing/test-agent-shutdown.sh]: Agreed - fixed.

**Comment by fidencio** [comment]
https://github.com/kata-containers/kata-containers/issues/1808 has been merged but this one was not, which is causing issues with the CI.

I'm running the tests and I'll do a pass of review as soon as we ensure the CI is passing.

**Review by fidencio — APPROVED** [review]
@jodh-intel, this commit is heavy and demanding to review.

I've taken a look at this, it looks sane, and I will merge it (rather than reverting its kata-containers counterpart), but I really would like to have a small chat / session with you in the near future to ensure I get all the bits & pieces correctly (and sorry for voluteering you to do so).

**Pull request merged**